Add quality declarations for each package except test_msgs

brawner opened the pull request:
This adds a QUALITY_DECLARATION.md for each package that describes aspirationally how it meets quality level 1. Aside from the dependent PRs (#89, #90, #91), this repo does not need further modifications to achieve QL 1 after an API review. 

There is one unresolved question regarding the message and service file definitions. It has not been historic practice to include a copyright statement in a message or service definition, which may lead to questions about who retains the copyright for such files considering many were copied over from non-copyrighted message/service file definitions in ROS 1. The QD rep generally requires copyright statements in source files, so this package would be atypical.

Furthermore, since none of the files in these QL 1 packages have copyrighted files, it's unclear what is even under the Apache 2 license and what purpose that license serves for this package.

New commit: https://github.com/ros2/rcl_interfaces/pull/92/commits/b2d5ca19755057eab649012cb4504c70d63c6fa5

brawner committed:
Create package-level READMEs and add some inline comments

Signed-off-by: Stephen Brawner <brawner@gmail.com>

brawner committed:
Add QUALITY_DECLARATION for all but test_msgs

Signed-off-by: Stephen Brawner <brawner@gmail.com>

Switch QDs to their QL current-level (4)

Signed-off-by: Stephen Brawner <brawner@gmail.com>

brawner commented:
This PR was updated to reflect the current status of each package (4).

brawner commented:
New commit: https://github.com/ros2/rcl_interfaces/pull/92/commits/b2d5ca19755057eab649012cb4504c70d63c6fa5

ahcorde commented on action_msgs/QUALITY_DECLARATION.md:
```diff
@@ -0,0 +1,109 @@
+
+The package `action_msgs` claims to be in the **Quality Level 4** category.
+
+Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Requirements for Quality Level 4 in REP-2004](https://www.ros.org/reps/rep-2004.html).
```
```suggestion
Below are the rationales, notes, and caveats for this claim, organized by each requirementlisted in the [Package Quality Categories in REP-2004](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#package-quality-categories) of the ROS2 developer guide.
```

brawner replied:
I'm currently linking to the non-existent rep page because the content in the developer guide is slated for removal.

ahcorde commented on action_msgs/QUALITY_DECLARATION.md:
```diff
@@ -0,0 +1,109 @@
+
+### License [3.iii]
+
+The license for `action_msgs` is Apache 2.0, the type is declared in the `package.xml` manifest file, and a full copy of the license is in the repository level [`LICENSE`](../LICENSE) file.
```
```suggestion
The license for `action_msgs` is Apache 2.0, the type is declared in the [package.xml](./package.xml) manifest file, and a full copy of the license is in the repository level [`LICENSE`](../LICENSE) file.
```

brawner replied:
Done

ahcorde commented on action_msgs/QUALITY_DECLARATION.md:
```diff
@@ -0,0 +1,109 @@
+
+The license for `action_msgs` is Apache 2.0, the type is declared in the `package.xml` manifest file, and a full copy of the license is in the repository level [`LICENSE`](../LICENSE) file.
+
+There is an automated test which runs a linter that ensures each file has a license statement.
```
link Link http://build.ros2.org/view/Epr/job/Epr__rcl_interfaces__ubuntu_bionic_amd64/lastBuild/

brawner replied:
Done

ahcorde commented on action_msgs/README.md:
```diff
@@ -0,0 +1,16 @@
+* [CancelGoal](srv/CancelGoal.srv): Cancel Goals either by id and/or timestamp
+
+## Quality Declaration
+This package claims to be in the **Quality Level 1** category, see the [Quality Declaration](QUALITY_DECLARATION.md) for more details.
```
```suggestion
This package claims to be in the **Quality Level 4** category, see the [Quality Declaration](QUALITY_DECLARATION.md) for more details.
```

brawner replied:
Done

ahcorde reviewed (changes requested):
I reviewed action_msgs QD. The same comments can be applied in the others QD. 

Please all review this other PR https://github.com/ros2/rcl_interfaces/pull/91 which is pending to merge this PR

brawner committed:
PR Fixup

Signed-off-by: Stephen Brawner <brawner@gmail.com>

brawner committed:
PR Fixup

Signed-off-by: Stephen Brawner <brawner@gmail.com>

brawner commented:
Ok, I addressed the feedback and applied it to the rest of the quality declarations. This PR has become a bit tricky to review since it depends on #91 and it includes two commits itself. I can squash the two commits here if that would make it easier to review. Otherwise you can see an overall comparison of just this PR here:
https://github.com/brawner/rcl_interfaces/compare/comments_documentation...brawner:quality_declaration

ahcorde commented on action_msgs/QUALITY_DECLARATION.md:
```diff
@@ -0,0 +1,111 @@
+
+### Peer Review Policy [2.iii]
+
+ Following the recommended guidelines for ROS Core packages, all pull requests must have at least 1 peer review.
```
nit: Remove the whitespace at the beginning. In the other QDs as well

ahcorde commented on lifecycle_msgs/msg/State.msg:
```diff
@@ -1,26 +1,57 @@
-# Primary state definitions
-# as depicted in
+# Primary state definitions as depicted in
```
```suggestion
# Primary state definitions as depicted in.
```

brawner replied:
Addressed in https://github.com/ros2/rcl_interfaces/pull/91

ahcorde commented on lifecycle_msgs/msg/State.msg:
```diff
@@ -1,26 +1,57 @@
+# Indicates state has not yet been set
 uint8 PRIMARY_STATE_UNKNOWN = 0
+
+# This is the life cycle state the node is in immediately after being
```
```suggestion
# This is the lifecycle state the node is in immediately after being
```

brawner replied:
Life cycle is two words, not sure why the package name uses one

ahcorde commented on lifecycle_msgs/msg/State.msg:
```diff
@@ -1,26 +1,57 @@
+# This state represents a node that is not currently performing any processing.
 uint8 PRIMARY_STATE_INACTIVE = 2
+
+# This is the main state of the node’s life cycle. While in this state, the node
```
```suggestion
# This is the main state of the node’s lifecycle. While in this state, the node
```

brawner replied:
Same

ahcorde commented on lifecycle_msgs/README.md:
```diff
@@ -0,0 +1,23 @@
+These messages and services form a standardized interface for transitioning these
+managed nodes through a known state-machine.
+
+For more information about life cycle nodes see: [design.ros2.org](http://design.ros2.org/articles/node_lifecycle.html).
```
```suggestion
For more information about lifecycle nodes see: [design.ros2.org](http://design.ros2.org/articles/node_lifecycle.html).
```

brawner replied:
Same

ahcorde commented on lifecycle_msgs/msg/State.msg:
```diff
@@ -1,26 +1,57 @@
+# data, produces output, etc.
 uint8 PRIMARY_STATE_ACTIVE = 3
+
+# The Finalized state is the state in which the node ends immediately before
```
punctuation

ahcorde commented on lifecycle_msgs/msg/Transition.msg:
```diff
@@ -1,25 +1,49 @@
-# Default values for transitions
-# as described on
+# Default values for transitions as described on
```
nit: punctuation

brawner replied:
Addressed in https://github.com/ros2/rcl_interfaces/pull/91

ahcorde commented on lifecycle_msgs/msg/TransitionDescription.msg:
```diff
@@ -1,3 +1,8 @@
+# The transition id and label of this description
```
punctuation

brawner replied:
Addressed in https://github.com/ros2/rcl_interfaces/pull/91

ahcorde reviewed (changes requested):
There are some files that require punctuation fixes and other minor fixes

ahcorde commented on action_msgs/QUALITY_DECLARATION.md:
```diff
@@ -0,0 +1,111 @@
+
+### ABI Stability Within a Released ROS Distribution [1.v]/[1.vi]
+
+`action_msgs` does not contain any c or c++ code and therefore will not affect ABI stability.
```
`C` and `C++` in uppercase? same in the other QDs

brawner replied:
Done

brawner committed:
PR Fixup

Signed-off-by: Stephen Brawner <brawner@gmail.com>

brawner committed:
PR fixup

Signed-off-by: Stephen Brawner <brawner@gmail.com>

ahcorde commented on composition_interfaces/README.md:
```diff
@@ -0,0 +1,13 @@
+## Services (.srv)
+* [ListNodes](srv/ListNodes.srv): Retrieve a list of running composable nodes, including their names and ids.
+* [LoadNodes](srv/LoadNode.srv): Load a composable node.
+* [UnloadNode](UnloadNode.srv): Unload a specified node by its id.
```
```suggestion
* [UnloadNode](srv/UnloadNode.srv): Unload a specified node by its id.
```

brawner replied:
Done.

ahcorde commented on composition_interfaces/srv/LoadNode.srv:
```diff
@@ -1,28 +1,42 @@
+# key/value arguments that are specific to a type of container process.
 rcl_interfaces/Parameter[] extra_arguments
 ---
 # True if the node was successfully loaded
```
```suggestion
# True if the node was successfully loaded.
```

brawner replied:
Done.

ahcorde commented on lifecycle_msgs/msg/State.msg:
```diff
@@ -1,26 +1,57 @@
+# These are the primary states. State changes can only be requested when the
+# node is in one of these states.
+
+# Indicates state has not yet been set
```
```suggestion
# Indicates state has not yet been set.
```

brawner replied:
Done.

ahcorde commented on lifecycle_msgs/msg/State.msg:
```diff
@@ -1,26 +1,57 @@
 uint8 TRANSITION_STATE_CLEANINGUP = 11
+
+# In this transition state the callback onShutdown will be executed to do any
+# cleanup necessary before destruction
```
punctation here and some more below

brawner replied:
Done.

ahcorde commented on lifecycle_msgs/srv/ChangeState.srv:
```diff
@@ -1,3 +1,7 @@
 Transition transition
 ---
+
+# Indicates whether the service was able to initiate the state transition
```
```suggestion
# The requested transition. This change state service will fail if the
# transition is not possible.
Transition transition
---

# Indicates whether the service was able to initiate the state transition.
```

brawner replied:
Done.

ahcorde commented on lifecycle_msgs/srv/GetAvailableStates.srv:
```diff
@@ -1,2 +1,3 @@
 ---
+# Array of possible states that can be transitioned to
```
```suggestion
# Array of possible states that can be transitioned to.
```

brawner replied:
Done.

ahcorde commented on lifecycle_msgs/srv/GetAvailableTransitions.srv:
```diff
@@ -1,2 +1,3 @@
 ---
+# An array of the possible start_state-goal_state transitions
```
```suggestion
# An array of the possible start_state-goal_state transitions.
```

brawner replied:
Done.

ahcorde commented on lifecycle_msgs/srv/GetState.srv:
```diff
@@ -1,2 +1,3 @@
 ---
+# The current state-machine state of the node
```
```suggestion
# The current state-machine state of the node.
```

brawner replied:
Done.

ahcorde commented on action_msgs/msg/GoalStatus.msg:
```diff
@@ -1,15 +1,32 @@
+# The goal was terminated by the action server without an external request.
 int8 STATUS_ABORTED   = 6
 
 # Goal info (contains ID and timestamp)
```
```suggestion
# Goal info (contains ID and timestamp).
```

brawner replied:
Done.

ahcorde commented on action_msgs/README.md:
```diff
@@ -0,0 +1,16 @@
+For more information about ROS 2 interfaces, see [index.ros2.org](https://index.ros.org/doc/ros2/Concepts/About-ROS-Interfaces/)
+
+## Messages (.msg)
+* [GoalInfo](msg/GoalInfo.msg): Goal identifier message, with a goal id and time stamp
+* [GoalStatus](msg/GoalStatus.msg): Describes a goal's current state machine status
+* [GoalStatusArray](msg/GoalStatusArray.msg): An array of [GoalStatus](msg/GoalStatus.msg) messages
```
```suggestion
* [GoalInfo](msg/GoalInfo.msg): Goal identifier message, with a goal id and time stamp.
* [GoalStatus](msg/GoalStatus.msg): Describes a goal's current state machine status.
* [GoalStatusArray](msg/GoalStatusArray.msg): An array of [GoalStatus](msg/GoalStatus.msg) messages.
```

brawner replied:
Done.

ahcorde reviewed (changes requested):
Same punctuation corrections and a link

brawner committed:
PR Fixup

Signed-off-by: Stephen Brawner <brawner@gmail.com>

brawner reviewed (commented):
Thanks, I believe I addressed your feedback.

brawner commented:
Verifying all ros2 packages build, and I didn't break anything.

[![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10575)](https://ci.ros2.org/job/ci_linux/10575/)

brawner commented:
CMake warning is unrelated to this documentation.
